k8s, node: Restore router IPs (cilium_host) from K8s resource

christarazi opened the pull request:
Previously, after a node reboot, Cilium would allocate a new router IP
and append it slice of node IPs. Since the node IPs have already been
synced to the K8s resource, meaning there are already IPs present (from
the previous Cilium instance), the router IP is appended to the slice.
In other parts of Cilium, it is assumed that the router IP is the first
node IP (first element of the slice). Since the new router IP has been
appended to the end, it is no longer where it is expected, aka no longer
the first element. This causes a mismatch of which router IP is to be
used. There should only ever be one router IP (one IPv4 or one IPv6).

In case of a node reboot, the router IPs cannot be restored because they
are wiped away due to the Cilium state dir being mounted as a tmpfs [1].
This commit fixes this to restore the router IPs from the K8s resource
(Node or CiliumNode) if they are present in the annotations. This
prevents the possibility of having more than one router IP, as described
above. Note that router IPs from the K8s resource are only restored if
no router IP was found on the filesystem, which is considered the source
of truth. In other words, the filesystem takes precedence over the K8s
resource. The user is warned in cases of a mismatch between the two
different sources. We also check that the IP to be restored is within
the pod / node CIDR range, otherwise we ignore it from restoration.

[1]: Linux distributions mount /run as tmpfs and Cilium's default state
	 directory is created under /run. (It's worth mentioning that it's
	 also common for /var/run to be symlinked to /run.)

Fixes: https://github.com/cilium/cilium/issues/16279

```release-note
Fix bug where Cilium allocates a new router (`cilium_host`) IP upon node reboot, breaking connectivity especially with IPsec
```

christarazi committed:
k8s, node: Restore router IPs (`cilium_host`) from K8s resource

Previously, after a node reboot, Cilium would allocate a new router IP
and append it slice of node IPs. Since the node IPs have already been
synced to the K8s resource, meaning there are already IPs present (from
the previous Cilium instance), the router IP is appended to the slice.
In other parts of Cilium, it is assumed that the router IP is the first
node IP (first element of the slice). Since the new router IP has been
appended to the end, it is no longer where it is expected, aka no longer
the first element. This causes a mismatch of which router IP is to be
used. There should only ever be one router IP (one IPv4 or one IPv6).

In case of a node reboot, the router IPs cannot be restored because they
are wiped away due to the Cilium state dir being mounted as a tmpfs [1].
This commit fixes this to restore the router IPs from the K8s resource
(Node or CiliumNode) if they are present in the annotations. This
prevents the possibility of having more than one router IP, as described
above. Note that router IPs from the K8s resource are only restored if
no router IP was found on the filesystem, which is considered the source
of truth. In other words, the filesystem takes precedence over the K8s
resource. The user is warned in cases of a mismatch between the two
different sources. We also check that the IP to be restored is within
the pod / node CIDR range, otherwise we ignore it from restoration.

[1]: Linux distributions mount /run as tmpfs and Cilium's default state
     directory is created under /run. (It's worth mentioning that it's
     also common for /var/run to be symlinked to /run.)

Fixes: https://github.com/cilium/cilium/issues/16279

Signed-off-by: Chris Tarazi <chris@isovalent.com>

nathanjsweet reviewed (commented):
I can't think of anything important offhand, but I guess there is the possibility that some kind of fast restart with a swapped NIC in a cloud provider could turn this into a race condition. However, that's a very edge case.

brb reviewed (commented):
What's going to happen if podCIDR will change? Will it still try to restore potentially obsolete router IP?

aanm commented on pkg/k8s/init.go:
```diff
@@ -267,3 +269,22 @@ func WaitForNodeInformation() error {
 }
+
+func restoreRouterHostIPs(n *nodeTypes.Node) {
+	if option.Config.EnableHostIPRestore {
```
nit: negate this to do an early return

aanm reviewed (changes requested):
>> What's going to happen if podCIDR will change? Will it still try to restore potentially obsolete router IP?
>
> @brb Do we support pod CIDRs changing anyway?

Well, it's not up to Cilium as it will simply receive the podCIDR from Kubernetes.

There is at least one case where this patch might cause issues.

1. Stop kubelet
2. Delete node with `kubectl delete node <node-name>`
3. Restart kubelet
4. Kubernetes will assign a new podCIDR to that node
5. Cilium will wrongly restore an IP address that does not belong to the podCIDR.

christarazi committed:
node: Modify SetIPv6NodeRange() to accept cidr.CIDR

This conforms SetIPv6NodeRange() to have the same prototype as
SetIPv4AllocRange(). There was no benefit for them to be different. It
will ease the subsequent commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
node: Clear router IPs on Uninitialize()

The subsequent commit will add unit tests that make use of
ipv{4,6}RouterAddress and state will need to be cleared during testing.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
